src/block_neural.rs

```diff
@@ -293,159 +268,76 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockNeuronLayer<L> {
-                        .fill(0.0);
-
-                    let (input_tape, output_tape) = block_helpers::get_input_output_borrows(
-                        &mut pb.tape,
-                        self.input_offset,
-                        self.num_inputs,
-                        self.output_offset,
-                        self.num_neurons,
-                    );
-
-                    for j in 0..self.num_neurons as usize {
-                        if self.dropout != 0.0
-                            && *self.rng_scratchpad.get_unchecked(j) < self.dropout_threshold
-                        {
-                            continue;
-                        }
-
-                        let general_gradient = output_tape.get_unchecked(j) * self.dropout_inv;
 
-                        let j_offset = j * self.num_inputs as usize;
+                    if self.max_norm != 0.0 && fb.example_number % 10 == 0 {
+                        let mut wsquaredsum: f32 = 0.000001; // Epsilon
                         for i in 0..self.num_inputs as usize {
-                            let feature_value = input_tape.get_unchecked(i);
-                            let gradient = general_gradient * feature_value;
-                            let update = self.optimizer.calculate_update(
-                                gradient,
-                                &mut self
-                                    .weights_optimizer
-                                    .get_unchecked_mut(i + j_offset)
-                                    .optimizer_data,
-                            );
-                            *output_errors.get_unchecked_mut(i) +=
-                                self.weights.get_unchecked(i + j_offset).weight * general_gradient;
-                            self.weights.get_unchecked_mut(i + j_offset).weight -= update;
+                            let w = *self.weights.get_unchecked(i + j_offset);
+                            wsquaredsum += w * w;
                         }
-                        {
-                            // Updating bias term:
-                            let gradient = general_gradient * 1.0;
-                            let update = self.optimizer.calculate_update(
-                                gradient,
-                                &mut self
-                                    .weights_optimizer
-                                    .get_unchecked_mut(bias_offset + j)
-                                    .optimizer_data,
-                            );
-                            self.weights.get_unchecked_mut(bias_offset + j).weight -= update;
-                        }
-
-                        if self.max_norm != 0.0 && fb.example_number % 10 == 0 {
-                            let mut wsquaredsum = 0.000001; // Epsilon
+                        let norm = wsquaredsum.sqrt();
+                        if norm > self.max_norm {
+                            let scaling = self.max_norm / norm;
```

Comment:
if there are any divisions in "hot" parts remaining, perhaps worth considering mult+shift transforms, as e.g., https://docs.rs/strength_reduce/latest/strength_reduce/